### PR TITLE
Add `NetworkKind`

### PR DESCRIPTION
**Pull request by tcharding**
Add a new type `NetworkKind` the describes the kind of network we are on, ether mainnet or one of the test nets (testnet, regtest, signet).

Use it in `Address`, `PrivateKey`, and `bip32`.

### CODE REVIEW

File: bitcoin/src/network.rs [hunk 1]
```diff
@@ -31,6 +31,33 @@ use crate::constants::ChainHash;
...
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum NetworkKind {
+    /// The Bitcoin mainnet network.
+    Main,
+    /// Some kind of testnet network.
+    Test,
+}
+
+// We explicitly do not provide `is_testnet`, using `!network.is_mainnet()` is less
+// ambiguous due to confusion caused by signet/testnet/regtest.
```
Comment (Kixunil) [inline bitcoin/src/network.rs:44]: Could be `is_test` or `is_a_test` to resolve the confusion.

File: bitcoin/src/address/mod.rs [hunk 2]
```diff
@@ -412,18 +364,17 @@ impl Address {
     ///
     /// This is the preferred non-witness type address.
     #[inline]
-    pub fn p2pkh(pk: impl Into<PubkeyHash>, network: Network) -> Address {
+    pub fn p2pkh(pk: impl Into<PubkeyHash>, network: impl Into<NetworkKind>) -> Address {
```
Comment (Kixunil) [inline bitcoin/src/address/mod.rs:367]: Makes me a bit worried that if someone invents a new network and decides to also add address prefix the whole thing breaks.

File: bitcoin/src/address/mod.rs [hunk 3]
```diff
@@ -800,20 +755,20 @@ impl FromStr for Address<NetworkUnchecked> {
 
         let inner = match *prefix {
             PUBKEY_ADDRESS_PREFIX_MAIN => {
-                let hash = PubkeyHash::from_byte_array(data);
-                AddressInner::P2pkh { hash, prefix: LegacyP2pkhPrefix::Mainnet }
+                let hash = PubkeyHash::from_slice(&data).expect("length checked above");
```
Comment (Kixunil) [inline bitcoin/src/address/mod.rs:758]: Why? It was previously statically correct.

File: bitcoin/src/crypto/key.rs [hunk 4]
```diff
@@ -305,17 +308,16 @@ impl PrivateKey {
...
 
     /// Format the private key to WIF format.
+    #[rustfmt::skip]
     pub fn fmt_wif(&self, fmt: &mut dyn fmt::Write) -> fmt::Result {
         let mut ret = [0; 34];
-        ret[0] = match self.network {
-            Network::Bitcoin => 128,
-            Network::Testnet | Network::Signet | Network::Regtest => 239,
-        };
+        ret[0] = if self.network.is_mainnet() { 128 } else { 239 };
```
Comment (Kixunil) [inline bitcoin/src/crypto/key.rs:319]: I like `match` more but don't bother with it here.

File: bitcoin/src/network.rs (hunk 1 above)
Comment (apoelstra) [reply bitcoin/src/network.rs:44]: Those both sound good to me. Or `is_test_network`.

File: bitcoin/src/address/mod.rs (hunk 2 above)
Comment (apoelstra) [reply bitcoin/src/address/mod.rs:367]: It's hard to imagine that. p2pkh addresses have been basically deprecated for 8+ years now. And no existing test networks have changed this prefix.

Comment (Kixunil) [reply bitcoin/src/address/mod.rs:367]: That's why it's only a bit. :) I think this one is acceptable.

File: bitcoin/src/address/mod.rs (hunk 3 above)
Comment (tcharding) [reply bitcoin/src/address/mod.rs:758]: Yes, no idea what I was thinking when I did this. Fixed.

**Comment by apoelstra** [comment]
This looks great to me. Am a little worried about the bip32 case where theoretically there could be more prefixes added (and no doubt we'll get the ypub/zpub people whining that we're making it even harder for them to implement their weird layer violations). But I'm doubtful. The model we've moved to, in my view, is that keys are "just keys" and networks should be specified elsewhere, e.g. at the descriptor layer.

**Review by apoelstra — DISMISSED** [review]
ACK ee2f83aa5e09efacd9b7927ca59f10cb7d8ead17

**Comment by tcharding** [comment]
Rebased, no other changes.

**Review by Kixunil — APPROVED** [review]
ACK 4354f37f5183f13f1d7cb1fc57d5d276a4a568ed

**Review by apoelstra — APPROVED** [review]
ACK 4354f37f5183f13f1d7cb1fc57d5d276a4a568ed

**Comment by darosior** [comment]
This is missing the API break label.

**Pull request merged**